core/src/slot_toggled_field.rs

```diff
@@ -24,25 +32,109 @@ impl SlotToggle {
         self.slot_removed
     }
 
-    pub fn activate(&mut self, slot: u64) -> bool {
-        if self.slot_added >= self.slot_removed {
-            false
-        } else {
-            self.slot_added = slot;
-            true
+    pub fn activate(&mut self, slot: u64, epoch_length: u64) -> bool {
+        match self.state(slot, epoch_length) {
+            SlotToggleState::Inactive => {
+                self.slot_added = slot;
+                true
+            }
+            _ => false,
+        }
+    }
+
+    pub fn deactivate(&mut self, slot: u64, epoch_length: u64) -> bool {
+        match self.state(slot, epoch_length) {
+            SlotToggleState::Active => {
+                self.slot_removed = slot;
+                true
+            }
+            _ => false,
         }
     }
 
-    pub fn deactivate(&mut self, slot: u64) -> bool {
-        if self.slot_added < self.slot_removed {
-            false
+    pub fn is_staked_or_slashable(&self, slot: u64, epoch_length: u64) -> bool {
+        matches!(
+            self.state(slot, epoch_length),
+            SlotToggleState::Active | SlotToggleState::Cooldown
+        )
+    }
+
+    pub fn is_active(&self, slot: u64, epoch_length: u64) -> bool {
```

Comment:
imo just have this return matches! State::Active

Reply:
Implemented what we talked about